apps/recnet-api/src/modules/rec/rec.service.ts

```diff
@@ -0,0 +1,259 @@
+        article.link
+      );
+      if (dbArticle) {
+        targetArticleId = dbArticle.id;
+      } else {
+        const newArticle = await this.articleRepository.createArticle(article);
+        targetArticleId = newArticle.id;
+      }
+    }
+
+    const articleIdToConnect = articleId ? articleId : targetArticleId;
+    if (articleIdToConnect === null) {
+      throw new RecnetError(
+        ErrorCode.INTERNAL_SERVER_ERROR,
+        HttpStatus.INTERNAL_SERVER_ERROR,
+        "Error getting article id while creating rec"
+      );
+    }
+    const newRec = await this.recRepository.createRec(
+      userId,
+      description,
+      articleIdToConnect
+    );
+    return {
+      rec: this.getRecFromDbRec(newRec),
+    };
+  }
+
+  public async updateUpcomingRec(
+    articleId: string | null,
+    article: CreateArticleInput | null,
+    description: string,
+    userId: string
+  ): Promise<UpdateRecResponse> {
+    const dbRec = await this.recRepository.findUpcomingRec(userId);
+    if (!dbRec) {
+      throw new RecnetError(
+        ErrorCode.INTERNAL_SERVER_ERROR,
+        HttpStatus.NOT_FOUND,
+        "Upcoming rec not found"
+      );
+    }
+    // article and articleId cannot be null at the same time
+    // check libs/recnet-api-model/src/lib/api/rec.ts
+    if (!article && !articleId) {
+      throw new RecnetError(
+        ErrorCode.REC_UPDATE_OR_CREATE_RULE_VIOLATION,
+        HttpStatus.BAD_REQUEST,
+        "Article and articleId cannot be null at the same time"
+      );
+    }
+
+    let targetArticleId: string | null = null;
+    if (article) {
+      if (articleId) {
+        throw new RecnetError(
+          ErrorCode.REC_UPDATE_OR_CREATE_RULE_VIOLATION,
+          HttpStatus.BAD_REQUEST,
+          "Article and articleId cannot have value at the same time"
+        );
+      }
+      // check if article already exists
+      const dbArticle = await this.articleRepository.findArticleByLink(
+        article.link
+      );
+      if (dbArticle) {
+        targetArticleId = dbArticle.id;
+      } else {
+        const newArticle = await this.articleRepository.createArticle(article);
+        targetArticleId = newArticle.id;
+      }
+    }
+
+    const articleIdToConnect = articleId ? articleId : targetArticleId;
+    if (articleIdToConnect === null) {
+      throw new RecnetError(
+        ErrorCode.INTERNAL_SERVER_ERROR,
+        HttpStatus.INTERNAL_SERVER_ERROR,
+        "Error getting article id while updating rec"
+      );
+    }
+    if (articleIdToConnect === dbRec.id) {
+      const updatedRec = await this.recRepository.updateRec(
+        dbRec.id,
+        description
+      );
+      return {
+        rec: this.getRecFromDbRec(updatedRec),
+      };
+    } else {
+      const updatedRec = await this.recRepository.updateRec(
+        dbRec.id,
+        description,
+        articleIdToConnect
+      );
+      return {
+        rec: this.getRecFromDbRec(updatedRec),
+      };
+    }
```

Comment:
I think articleIdToConnect will never equal to dbRec.id. It should probably be `articleIdToConnect === dbRec.article.id`. Also, I prefer to put the return syntax at the end of the method if possible.
```ts
    let updatedRec: DbRec;
    if (articleIdToConnect === dbRec.article.id) {
      updatedRec = await this.recRepository.updateRec(dbRec.id, description);
    } else {
      updatedRec = await this.recRepository.updateRec(
        dbRec.id,
        description,
        articleIdToConnect
      );
    }
    return {
      rec: this.getRecFromDbRec(updatedRec),
    };
```